crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -401,29 +401,72 @@ def upgrade_drbd_cluster(cluster)
           return true
         end
 
+        all_upgraded = true
+        drbd_nodes.each do |drbd_node|
+          all_upgraded &&= drbd_node.upgraded?
+        end
+        if all_upgraded
+          save_upgrade_state("All nodes in cluster #{cluster} have already been upgraded.")
+          return true
+        end
+
         # First node to upgrade is DRBD slave. There might be more resources using DRBD backend
         # but the Master/Slave distribution might be different among them.
         # Therefore, we decide only by looking at the first DRBD resource we find in the cluster.
-        drbd_slave = nil
-        drbd_master = nil
+        first = nil
+        second = nil
 
-        drbd_nodes.each do |drbd_node|
-          cmd = "LANG=C crm resource status ms-drbd-{postgresql,rabbitmq}\
-          | sort | head -n 2 | grep \\$(hostname) | grep -q Master"
-          out = drbd_node.run_ssh_cmd(cmd)
-          if out[:exit_code].zero?
-            drbd_master = drbd_node
-          else
-            drbd_slave = drbd_node
+        # But we have to make sure that when this method is invoked for a second time
+        # (probably after previous failure), same order of nodes is selected as in the first case.
+        # Looking at the DRBD state could be problematic, because master and slave are normally
+        # switched during the upgrade. So let's look at the data we have about the upgraded nodes.
+        status = ::Crowbar::UpgradeStatus.new
+        # current_node is only set when the upgrade of a node starts, and is empty at the start.
+        current_node = status.progress[:current_node] || {}
+        current_name = current_node[:name]
+        # upgraded_nodes number is increased only after a node upgrade is finished.
+        upgraded_nodes = status.progress[:upgraded_nodes] || 0
+
+        if current_name.nil?
+          # No node upgrade has started yet, so let's pick the first/second as slave/master
+          drbd_nodes.each do |drbd_node|
+            cmd = "LANG=C crm resource status ms-drbd-{postgresql,rabbitmq}\
+            | sort | head -n 2 | grep \\$(hostname) | grep -q Master"
+            out = drbd_node.run_ssh_cmd(cmd)
+            if out[:exit_code].zero?
+              second = drbd_node
+            else
+              # this is DRBD slave
+              first = drbd_node
+            end
+          end
+        else
+          # In this branch, we know that upgrade of DRBD nodes was already started but failed
+          # and the method was executed again. Let's find out which node went first/second initialy
+          # and stick to that order.
```

Comment:
Somehow I think we should add a big warning here (and maybe in the commit message as well) that this part of the code (i.e. the following ~15 lines) heavily depend on the fact that the drbd cluster is the first one to be upgraded. I think for now it's ok to depend on that, but we need to aware of it.